util/vole.go

```diff
@@ -0,0 +1,297 @@
+			requestedCids[c] = true
+			responded += 1
+		}
+	}
+
+loop:
+	for {
+		var res msgOrErr
+		select {
+		case res = <-rcv.result:
+		case <-ctx.Done():
+			break loop
+		}
+
+		if res.err != nil {
+			return nil, res.err
+		}
+
+		if res.msg == nil {
+			panic("should not be reachable")
+		}
+
+		for _, block := range res.msg.Blocks() {
+			markResponse(block.Cid())
+		}
+
+		for _, c := range res.msg.Haves() {
+			markResponse(c)
+		}
+
+		for _, c := range res.msg.DontHaves() {
+			markResponse(c)
+		}
+		if responded == len(requestedCids) {
+			return nil, nil
+		}
+	}
+
+	missingCids := make(map[cid.Cid]bool)
+	for c, responded := range requestedCids {
+		if !responded {
+			missingCids[c] = true
+		}
+	}
+	return missingCids, nil
+}
+
+func connectToBitSwap(ctx context.Context, targetPeer peer.ID, rcv *bsReceiver) (bsnet.BitSwapNetwork, error) {
+	// Connect randomly on either TCP or QUIC.
+	localhost := ""
+	if rand.Intn(2) == 1 {
+		localhost = "/ip4/127.0.0.1/udp/4001/quic/p2p/" + targetPeer.String()
+		bsLog.Infof("connecting through QUIC")
+	} else {
+		localhost = "/ip4/127.0.0.1/tcp/4001/p2p/" + targetPeer.String()
+		bsLog.Infof("connecting through TCP")
+	}
+	// FIXME: Any preference on either transport?
+	// FIXME: Do we want to make the transport an option to test separately
+	//  both TCP and QUIC (since the latter was suspected to be a culprit
+	//  here).
+	// FIXME: Check if the GW changes port number and where do we get it.
+
+	ma, err := multiaddr.NewMultiaddr(localhost)
+	if err != nil {
+		panic("failed to decode multi-address")
+	}
+
+	// FIXME: Can we reuse the `bitswapNetwork` in the IPFS core node
+	//  constructor? We would need to overwrite the receive delegate
+	//  (which is controlled by the BitSwap struct) or chain them
+	//  somehow. Right now we are re-connecting to ourselves at the tcp/libp2p
+	//  level creating a new host. (Maybe in terms of testing including this
+	//  layer, even at the bypassed localhost level, is useful.)
+	h, err := libp2p.New(ctx,
+		libp2p.Transport(tcp.NewTCPTransport),
+		libp2p.Transport(quic.NewTransport),
+
+		libp2p.Identity(privKey), // ID: 12D3KooWDpJ7As7BWAwRMfu1VU2WCqNjvq387JEYKDBj4kx6nXTN
+		// FIXME: Could we pass the node's key here? It would be the cleanest
+		//  but maybe BW is not prepared to handle a request from self.
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	ai, err := peer.AddrInfoFromP2pAddr(ma)
+	if err != nil {
+		return nil, err
+	}
+
+	hctx, _ := context.WithTimeout(ctx, time.Second*1) // Basic connection timeout for localhost.
+	if err := h.Connect(hctx, *ai); err != nil {
+		return nil, err
+	}
+
+	nilRouter, err := nrouting.ConstructNilRouting(nil, nil, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	bs := bsnet.NewFromIpfsHost(h, nilRouter)
+
+	bs.SetDelegate(rcv)
+
+	return bs, nil
+}
+
+func createCids(n int) map[cid.Cid]bool {
+	cids := make(map[cid.Cid]bool, 2*n)
+
+	// FIXME: ID hashes seem to be present by default,
```

Comment:
Potential issues:

1. Won't rule out datastore/blockstore issues.
2. Won't rule out issues where a node doesn't have a block then later gets a copy.

But we'll still get useful information this way.

Reply:
Agreed. Using ID and shortened SHA CIDs was the simplest way I found to trigger both found/not-found responses without actual manipulation of the store and pinning service. I'm open to other alternatives if we can code them in a few lines, otherwise we can ship this for the first iteration.